From a80c4964fd3e4834f492b300ad0a3a0543192ada Mon Sep 17 00:00:00 2001
From: AMD User <62472426+mqhc2020@users.noreply.github.com>
Date: Fri, 27 Jun 2025 09:09:12 +0000
Subject: [PATCH] Add functions for SGLang

---
 python/triton/runtime/cache.py | 12 ++++++++++++
 1 file changed, 12 insertions(+)

diff --git a/python/triton/runtime/cache.py b/python/triton/runtime/cache.py
index 7edb75cb8..d7db82912 100644
--- a/python/triton/runtime/cache.py
+++ b/python/triton/runtime/cache.py
@@ -3,11 +3,23 @@ import os
 import uuid
 from abc import ABC, abstractmethod
 from typing import Dict, List, Optional
+from pathlib import Path
 import base64
 import hashlib
 
 from .. import knobs
 
+def get_home_dir():
+    return os.getenv("TRITON_HOME", Path.home())
+
+def default_cache_dir():
+    return os.path.join(get_home_dir(), ".triton", "cache")
+
+def default_override_dir():
+    return os.path.join(get_home_dir(), ".triton", "override")
+
+def default_dump_dir():
+    return os.path.join(get_home_dir(), ".triton", "dump")
 
 class CacheManager(ABC):
 
-- 
2.34.1

